### PR TITLE
Staging branch for merging 5.10 kernel changes, gnupg2.4 and flashrom 1.3 (testing needed)

### PR DESCRIPTION
**Pull request by tlaurion**
This is a merger of #1422 #1381 and #1423 to facilitate testing. No point really asking board owners to test 3 PRs individually where they need to coexist to make things go forward and coexist in SPI space. This took a loooooong time (xx20/xx30 were still on 4.14 kernel) to happen. Hopefully, this will be easier in the future with the work that happened here. Its way easier to compare kernel configs and coreboot configs in oldconfig format, and helpers finally help comparing size differences of pull requests, and move from defconfig to oldconfig and see new kernel configuration options at version bumps to create diffs and apply to other board configs as patches.

---

THIS PR ALSO RENAMED THE FOLLOWING UNTESTED BOARDS https://github.com/osresearch/heads/compare/92e29c4..56a312e8edb91e773a0bbe0670af1574ca6ac274

Putting back boards from UNTESTED to normal will continue under #1421.
----

TLDR changeset:
- #1381 is an extensive effort to reduce and unify 5.10.5 kernel configs bumps in oldconfig format, reducing clutter for mostly all boards covered under Heads and finally permitting to compare board configs between each other.
  -  It also brings back libgfxinit, and shows a bootsplash of Heads logo at boot while the kernel is read from SPI and the iGPU drivers finally takes ownership of the framebuffer. Hope you like it.
- #1422 bumps gnupg toolstack to 2.4.0, needed to support new algos and ease further work. (thanks @saper)
- #1423 updates flashrom for newer boards, and improves the progress bar output to screen (thanks to @JonathonHall-Purism) @daringer you might want to look and test this.

----

Tested: oem-factory-reset, internal firmware upgrade, seal/unseal TOTP, sealing disk unlock key and kexec'ing into Qubes 4.2.

- [x] x230-maximized (@tlaurion)
- [x] x230-legacy (@tlaurion) 
  - [ ] Somebody uses legacy boards here? Please tag me, I would love to know. 
  - (note: that as documented per commit under #1381, network dev support is removed. There is a warning after kexec call saying the ifdown could not do its job. This can be neglected)
- [x] t430-legacy (thanks @nestire)
  - [x] Somebody uses legacy boards here? Please tag me, I would love to know. 
  - (note: that as documented per commit under #1381, network dev support is removed. There is a warning after kexec call saying the ifdown could not do its job. This can be neglected)
- [ ] x220 (xx20): @Thrilleratplay @BlackMaria @srgrint
  - [ ] Somebody x220 legacy board here (not x220-maximized)? Please tag me, I would love to know otherwise will be #1421 
- [x] [t420](https://doc.coreboot.org/mainboard/lenovo/t420.html) (xx20): Thanks @natterangell (iGPU)
- [x] [t430](https://doc.coreboot.org/mainboard/lenovo/t430.html) (xx30): thanks @nestire
- [ ] [t520](https://doc.coreboot.org/mainboard/lenovo/t520.html) (xx30):
  - [ ] Somebody owns this board? Please tag me, I would love to know otherwise will be #1421 
- [ ] [t530](https://doc.coreboot.org/mainboard/lenovo/t530.html) (xx30): @3hhh
  - Note: dGPU versions will not have libgfxinit. Would need iGPU/dGPU testing
- [ ] [w530](https://doc.coreboot.org/mainboard/lenovo/w530.html) (xx30): @eganonoa @zifxify @weyounsix (dGPU: w530-k2000m) @jnscmns (dGPU K1000M)
  - Note: dGPU versions will not have libgfxinit. Would need iGPU/dGPU testing
- [x] x230 (xx30): @tlaurion @osresearch @merge @jan23 @MrChromebox @shamen123 @eganonoa @bwachter @Thrilleratplay @jnscmns @doob85
  - [ ] Somebody uses legacy boards here? Please tag me, I would love to know otherwise will be #1421 
- [ ] X230i (x230): @natterangell
- [ ] x230-fhd/edp variant: @n4ru @computer-user123 @Tonux599 @househead @pcm720
- [x] [t440p](https://doc.coreboot.org/mainboard/lenovo/t440p.html): (thanks @srgrint!)
- [x] Librem 13v2 (Skylake): @JonathonHall-Purism
- [x] Librem 13v4 (Kabylake): @JonathonHall-Purism
- [x] Librem 15v3 (Skylake): @JonathonHall-Purism
- [x] Librem 15v4 (Kabylake): @JonathonHall-Purism
- [ ] kgpe-d16 AMD fam15h (dropped in coreboot 4.12): @0rb677 @Tonux599 @zifxify @blobless
  - [ ] Somebody owns and use this board here? Please tag me, I would love to know otherwise will be #1421  
- [x] [Talos II](https://docs.dasharo.com/variants/talos_2/overview/): @tlaurion
- [ ] [ASUS P8Z77 M PRO](https://doc.coreboot.org/mainboard/asus/p8z77-m_pro.html): @ThePlexus

Will add all boards touched by the merger so that board owners can test this instead of other PRs.

---

Also delays #1421 a little bit more.

### CODE REVIEW

**Review by tlaurion — COMMENTED** [review]
Don't forget to update crypto requirements before closing that issue

**Comment by pcm720** [comment]
Tested out [x230-maximized-fhd_edp CI build](https://app.circleci.com/pipelines/github/tlaurion/heads/1685/workflows/dd3186e9-d6e0-4ce6-b98b-e314b3b25ba1/jobs/22785/artifacts), everything seems to work fine.

**Comment by tlaurion** [comment]
@pcm720 this is awesome news, meaning that libgfxinit is not needed to apply vbt magic on coreboot side for the edp mod to work!!!

This means that https://github.com/osresearch/heads/pull/1398/files#diff-6d0b40c74d1c5ee9e26189f7420a2e8a61aef86dbafb3f846f238c8f5aad4bd3R395-R398
https://github.com/osresearch/heads/pull/1398/files#diff-6d0b40c74d1c5ee9e26189f7420a2e8a61aef86dbafb3f846f238c8f5aad4bd3R419-R420

And most importantly https://github.com/osresearch/heads/pull/1398/files#diff-6d0b40c74d1c5ee9e26189f7420a2e8a61aef86dbafb3f846f238c8f5aad4bd3R123 is still working as expected!!!!

I do not want to abuse of your time, but if you could confirm that https://github.com/osresearch/heads/pull/1381 works for your edp variant?

**Comment by srgrint** [comment]
Just adding another data point - flashed heads-t440p-maximized-v0.2.0-1571-g307f75d.rom on my t440p.

Seems to work fine - boots both debian installer from USB and installed version of debian from hard drive.  No obvious regressions.

**Comment by tlaurion** [comment]
Needs to be debased after #1381 is merged

**Comment by tlaurion** [comment]
Ok will stage changes from 
- #1381 [tlaurion:xx20_xx30_kernel_510_bump](https://github.com/tlaurion/heads/tree/xx20_xx30_kernel_510_bump)
- #1355 [saper:gnupg-1.4.23-udate](https://github.com/saper/heads/tree/gnupg-1.4.23-udate)
- #1251 [tlaurion:dasharo_flashrom_test](https://github.com/tlaurion/heads/tree/dasharo_flashrom_test)

**Comment by tlaurion** [comment]
Previous change didn't work since legacy boards do not have enough spi space.

Staging:
- #1381 [tlaurion:xx20_xx30_kernel_510_bump](https://github.com/tlaurion/heads/tree/xx20_xx30_kernel_510_bump)
- #1350 [saper:gnupg-2.4.0-update](https://github.com/saper/heads/tree/gnupg-2.4.0-update)
- #1251 [tlaurion:dasharo_flashrom_test](https://github.com/tlaurion/heads/tree/dasharo_flashrom_test)

Goal is to see if #1381 has liberated enough space for those to be merged as is later on.

**Comment by tlaurion** [comment]
Staging:

- #1381 
- #1422 
- #1423 

Goal is to see if https://github.com/osresearch/heads/pull/1381 has liberated enough space for those to be merged as is later on.

**Comment by tlaurion** [comment]
Images available through CircleCI for 30 days. Just click the green checkmark, select board, go into artifacts on the board CircleCI options, and download the full rom image for your board.

Of course, make sure the hash for the board ROM matches what is reported under BOARD output/hashes.txt. And flash internally as you do normally.

Please tag me directly if there is any issue/success. I will update OP accordingly.
Finally: this is moving forward and next steps can happen.

File: initrd/bin/flash.sh [hunk 1]
```diff
@@ -77,17 +68,43 @@ flashrom_progress() {
...
+            #    # "... is writable, writing range (0x000000..0x000fff)"
+            #    current=$(echo "$IN" | sed -nE 's/^\(0x[0-9a-f]+..(0x[0-9a-f]+)\)\.$/\1/p')
+            #    if [ "$current" != "" ]; then
+            #        percent=$((100 * (current + 1) / total_bytes))
+            #        echo "current: $current, total: $total_bytes, percent: $percent"
+            #        print_flashing_progress "$percent"
+            #    else
+            #        echo "failed to match: $IN"
+            #    fi
+            #fi
```
Comment (JonathonHall-Purism) [inline initrd/bin/flash.sh:85]: @tlaurion Could you delete this commented-out code?  Sorry this came from my WIP branch.  (Keep the walk_eraseblocks() notes below though please.)

Comment (tlaurion) [reply initrd/bin/flash.sh:85]: Done under #1423, will rebase when CircleCI actually builds a cache that it can reuse.

**Comment by tlaurion** [comment]
And here we go, staging #1381 #1422 and #1423 where those PR now have metrics to be reviewed based on size.

**Comment by tlaurion** [comment]
Alright!

Here are diff of sizes compared to master, explained in more details under https://github.com/osresearch/heads/pull/1381 https://github.com/osresearch/heads/pull/1422 and https://github.com/osresearch/heads/pull/1423
```
user@heads-tests-deb12:~/heads$ wget "https://circleci.com/api/v1.1/project/gh/osresearch/heads/10347/output/105/0?file=true&allocation-id=649b06dff63182304a39a9bb-0-build%2F9ED89B0" -O master -q
user@heads-tests-deb12:~/heads$ wget "https://circleci.com/api/v1.1/project/github/tlaurion/heads/25511/output/105/0?file=true&allocation-id=649b2878f3a6663e16e8b3a8-0-build%2F55310BE4" -O 1398 -q
user@heads-tests-deb12:~/heads$ diff -u master 1398
--- master	2023-06-27 15:19:51.433000000 -0400
+++ 1398	2023-06-27 15:19:56.470000000 -0400
@@ -1,23 +1,23 @@
-2023-06-27 11:59:09-04:00 91f65bed793c0165203b1cbb9ac4f390f15ffa0e clean
- 3039696:/root/project/build/x86/t520-hotp-maximized/bzImage
-  684032:/root/project/build/x86/t520-hotp-maximized/modules.cpio
------
-  304272:./lib/modules/e1000e.ko
-   64800:./lib/modules/ehci-hcd.ko
-   10728:./lib/modules/ehci-pci.ko
-  159840:./lib/modules/xhci-hcd.ko
-   10568:./lib/modules/xhci-pci.ko
-  132360:./lib/modules/usb-storage.ko
+2023-06-27 14:25:07-04:00 1a607830d5e2e382d8119b0e498cb2ebb6327f70 clean
+ 3055264:/root/project/build/x86/t520-hotp-maximized/bzImage
+  685056:/root/project/build/x86/t520-hotp-maximized/modules.cpio
+-----
+  304040:./lib/modules/e1000e.ko
+   63272:./lib/modules/ehci-hcd.ko
+   10896:./lib/modules/ehci-pci.ko
+  165024:./lib/modules/xhci-hcd.ko
+   11544:./lib/modules/xhci-pci.ko
+  128632:./lib/modules/usb-storage.ko
 -----
-10793984:/root/project/build/x86/t520-hotp-maximized/tools.cpio
+10828800:/root/project/build/x86/t520-hotp-maximized/tools.cpio
 -----
   596544:./lib/libc.so
   491024:./lib/libcairo.so.2
   138520:./lib/libcryptsetup.so.4
    64448:./lib/libassuan.so.0
- 1054216:./lib/libgcrypt.so.20
-  109520:./lib/libgpg-error.so.0
-  199016:./lib/libksba.so.8
+  776328:./lib/libgcrypt.so.20
+  121704:./lib/libgpg-error.so.0
+  207216:./lib/libksba.so.8
   171032:./lib/libpng16.so.16
    84584:./lib/libusb-1.0.so.0
    18440:./lib/libusb-0.1.so.4
@@ -40,13 +40,13 @@
     5904:./bin/qrenc
   118424:./bin/tpm
    73600:./bin/lspci
-  580568:./bin/flashrom
+  662264:./bin/flashrom
    50952:./bin/cryptsetup
    58200:./bin/cryptsetup-reencrypt
    28432:./bin/veritysetup
-  751456:./bin/gpg
-  314584:./bin/gpg-agent
-  346376:./bin/scdaemon
+  828136:./bin/gpg
+  343984:./bin/gpg-agent
+  450792:./bin/scdaemon
    44008:./bin/pinentry-tty
   121080:./bin/dmsetup
  1414928:./bin/lvm
@@ -65,7 +65,7 @@
   652688:./bin/bash
    10064:./bin/io386
    35432:./bin/cbmem
-     699:./etc/config
+     698:./etc/config
 -----
   239104:/root/project/build/x86/t520-hotp-maximized/heads.cpio
 -----
@@ -76,7 +76,7 @@
      237:./bin/cbfs.sh
     5557:./bin/config-gui.sh
     2612:./bin/flash-gui.sh
-    6653:./bin/flash.sh
+    6990:./bin/flash.sh
      411:./bin/flashrom-kgpe-d16-openbmc.sh
     1345:./bin/generic-init
     9056:./bin/gpg-gui.sh
@@ -98,7 +98,7 @@
     2511:./bin/media-scan
     2662:./bin/mount-usb
     1639:./bin/network-init-recovery
-   25179:./bin/oem-factory-reset
+   25034:./bin/oem-factory-reset
     2345:./bin/oem-system-info-xx30
      324:./bin/poweroff
      597:./bin/qubes-measure-luks
@@ -115,6 +115,7 @@
      380:./bin/wget-measure.sh
      639:./bin/xx30-flash.init
     3166:./etc/ash_functions
+      17:./etc/distro/gpg-agent.conf
     1725:./etc/distro/keys/archlinux.key
     1629:./etc/distro/keys/qubes-4.key
     3139:./etc/distro/keys/qubes-r4.1-weekly-builds-signing-key.asc
@@ -135,5 +136,5 @@
      924:./sbin/config-dhcp.sh
     1064:./sbin/insmod
 -----
- 3977216:build/x86/t520-hotp-maximized/initrd.cpio.xz
- 8388608:/root/project/build/x86/t520-hotp-maximized/heads-t520-hotp-maximized-v0.2.0-1577-g91f65be.rom
+ 4058112:build/x86/t520-hotp-maximized/initrd.cpio.xz
+ 8388608:/root/project/build/x86/t520-hotp-maximized/heads-t520-hotp-maximized-v0.2.0-1612-g1a60783.rom
```


-----

Using it on main laptop (x230-hotp-maximized). No regression into flashing clean, importing public key, sealing totp/hotp, signing /boot, setting new boot option, setting disk unlock key, and booting into Q4.1 with the chipset being finalized as expected. Past PR were test individually for flashing and gnupg while doing oem-factory-reset with a nitrokey pro.

libgfxinit bootpslash in msec after power button pushed (thanks @ThePlexus!):
signal-2023-06-27-150451_002

Reading flash while injecting pubkey:
signal-2023-06-27-150451_003

Writing flash after injecting pubkey:
signal-2023-06-27-150451_004

Bootting Q4.1 with TPM unsealing disk unlock key (finalizing chipset):
signal-2023-06-27-150451_005


Please test ASAP people. I would love to merge this!

**Comment by JonathonHall-Purism** [comment]
Reviewed the changes; all the changes affecting Librems are common (not model-specific).  Specifically compared the librem-common Linux config to master.  Discussed several changes with @tlaurion and I agree with everything except potentially CPU_ISOLATION=y.

I don't think we need CPU_ISOLATION=y, that's for a real-time task to run uninterrupted on one or more CPUs.  cryptsetup reencrypt was mentioned as a time consuming task for us, but I do not find anything suggesting that it supports task isolation.

I did functional testing on Librem 14, tested OEM reset, flash, boot, etc., everything is working fine.

Tested boot and reflash on 13v2, 13v4, 15v3, 15v4 (SKL/KBL boards).  Also tested OEM reset on 13v2 and 13v4.  Everything tested is working.

**Comment by tlaurion** [comment]
> Reviewed the changes; all the changes affecting Librems are common (not model-specific). Specifically compared the librem-common Linux config to master. Discussed several changes with @tlaurion and I agree with everything except potentially CPU_ISOLATION=y.

```
user@heads-tests-deb12:~/heads$ diff -u 1398 no_isolation 
--- 1398	2023-06-27 15:19:56.470000000 -0400
+++ no_isolation	2023-06-28 12:13:36.348000000 -0400
@@ -1,5 +1,5 @@
-2023-06-27 14:25:07-04:00 1a607830d5e2e382d8119b0e498cb2ebb6327f70 clean
- 3055264:/root/project/build/x86/t520-hotp-maximized/bzImage
+2023-06-28 10:30:12-04:00 0ffb8cd5150d9cece8d914965b1f8e9a10c3c606 clean
+ 3053888:/root/project/build/x86/t520-hotp-maximized/bzImage
   685056:/root/project/build/x86/t520-hotp-maximized/modules.cpio
 -----
   304040:./lib/modules/e1000e.ko
@@ -137,4 +137,4 @@
     1064:./sbin/insmod
 -----
  4058112:build/x86/t520-hotp-maximized/initrd.cpio.xz
- 8388608:/root/project/build/x86/t520-hotp-maximized/heads-t520-hotp-maximized-v0.2.0-1612-g1a60783.rom
+ 8388608:/root/project/build/x86/t520-hotp-maximized/heads-t520-hotp-maximized-v0.2.0-1613-g0ffb8cd.rom
```

@JonathonHall-Purism : small gains in size removing CPU_ISOLATION.
Will edit with results from cryptsetup-reencrypt tests but as of now, no real gain observed with it in.

**Comment by JonathonHall-Purism** [comment]
Thanks @tlaurion, with CPU_ISOLATION removed again I am good to merge this once the other boards are tested!

**Comment by srgrint** [comment]
Tested t440p-maximized/heads-t440p-maximized-v0.2.0-1613-g92e29c4.rom on my t440p

Seems to work fine - test included booting into Debian instead of Qubes as my machine dosen't have enough RAM for the latter.

**Comment by natterangell** [comment]
Tested heads-t420-hotp-maximized-v0.2.0-1613-g92e29c4.rom and everything seems to be working perfectly. Nice bootsplash 😊

**Comment by nestire** [comment]
tested heads-t430-hotp-maximized-v0.2.0-1613-g92e29c4.rom -> worked 

tested heads-t430-hotp-legacy-v0.2.0-1613-g92e29c4.rom:
- splash image -> works
- this "finalizing chipset" does not show up before booting Qubes 4.1 
- firmware upgrade/oem reset/ Booting Qubes -> works

**Comment by tlaurion** [comment]
A decision has been taken.
The following boards have been renamed to UNTESTED_* per [1576381](https://github.com/osresearch/heads/pull/1398/commits/1576381f04a7f8cff30a9b4422dafca9c649890c) since nobody tested this PR and previous ones.

Please participate under #1421 or open an issue saying the following boards have been tested and working on daily usage to revive them individually: 
- [t520](https://doc.coreboot.org/mainboard/lenovo/t520.html) (xx30):
  - Somebody owns this board? Please tag me, I would love to know otherwise will be #1421 
- [t530](https://doc.coreboot.org/mainboard/lenovo/t530.html) (xx30): @3hhh
  - Note: dGPU versions will not have libgfxinit. Would need iGPU/dGPU testing
- [w530](https://doc.coreboot.org/mainboard/lenovo/w530.html) (xx30): @eganonoa @zifxify @weyounsix (dGPU: w530-k2000m) @jnscmns (dGPU K1000M)
  - Note: dGPU versions will not have libgfxinit. Would need iGPU/dGPU testing
- x230-fhd/edp variant: @n4ru @computer-user123 @Tonux599 @househead @pcm720
- kgpe-d16 AMD fam15h (dropped in coreboot 4.12): @0rb677 @Tonux599 @zifxify @blobless
  - Somebody owns and use this board here? Please tag me, I would love to know otherwise will be #1421  
- [ASUS P8Z77 M PRO](https://doc.coreboot.org/mainboard/asus/p8z77-m_pro.html): @ThePlexus

Waiting for all boards to build successfully prior of merging.
Replaces https://github.com/osresearch/heads/pull/1381 https://github.com/osresearch/heads/pull/1422 and https://github.com/osresearch/heads/pull/1423

**Comment by tlaurion** [comment]
Board owners reported testing 92e29c4, not [a51143a](https://github.com/osresearch/heads/pull/1398/commits/a51143a6eb12b189a8cfb6f8d767a7dce2f8d38f) which is here by error. Rebasing.

**Comment by srgrint** [comment]
Have tested heads-x220-maximized-v0.2.0-1613-g92e29c4.rom on my x220.  No problems.

**Comment by tlaurion** [comment]
> Have tested heads-x220-maximized-v0.2.0-1613-g92e29c4.rom on my x220. No problems.

Just remove x220-hotp-maximized and x220-maximized boards from UNTESTED.
Also updating #1421

**Comment by srgrint** [comment]
Shouldn't t440p also be removed from UNTESTED?  I tested it a few days ago.

**Comment by tlaurion** [comment]
> Shouldn't t440p also be removed from UNTESTED? I tested it a few days ago.
Sorry about that, human error.

Readded t440p-maximized and t440p-hotp-maximized.  Will update #1421 as well.
Thanks @srgrint for flagging.

**Comment by tlaurion** [comment]
Other boards will be revived by other PR. Waiting for this to successfully build prior of merging.
Thanks all.

**Comment by tlaurion** [comment]
Other boards specified under #1421 OP wlll be revived by other individual  raised issues and cooresponding PRs. 

Waiting for [f8cb3db](https://github.com/osresearch/heads/pull/1398/commits/f8cb3db775059132f468f612ec095b5fc4980f01) this to successfully build prior of merging.

Thanks all.


----

Sorry for the noise, if you want to inspect final changes prior of staging PR and final PR that would be here:
https://github.com/osresearch/heads/compare/92e29c4..f8cb3db
